File: tool/knowledge_memory_v2.py
#!/usr/bin/env python3
"""Deterministic v1.8 knowledge, memory, skill, and object-store helpers."""
from __future__ import annotations

from dataclasses import dataclass, field
import datetime as dt
import hashlib
import json
from pathlib import Path
import re

VERSION = "1.9.0+190"

ALLOWED_SKILL_TOOLS = {
    "read_file",
    "inspect_file",
    "write_file",
    "run_command",
    "verify_project",
}


def canonical_json(value: object) -> str:
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def sha256_text(value: str) -> str:
    return sha256_hex(value.encode("utf-8"))


def utc_now() -> str:
    return dt.datetime.now(dt.timezone.utc).isoformat()


def _tokenize(value: str) -> list[str]:
    return re.findall(r"[a-z0-9_./-]+", value.lower())


@dataclass(frozen=True)
class StoredObject:
    sha256: str
    relative_path: str
    media_type: str
    size_bytes: int
    created_at: str
    labels: dict[str, str] = field(default_factory=dict)

    def to_json(self) -> dict[str, object]:
        return {
            "schemaVersion": "1.0.0",
            "sha256": self.sha256,
            "relativePath": self.relative_path,
            "mediaType": self.media_type,
            "sizeBytes": self.size_bytes,
            "createdAt": self.created_at,
            "labels": dict(sorted(self.labels.items())),
        }


class ObjectStore:
    def __init__(self, root: Path) -> None:
        self.root = root

    def relative_path(self, sha256: str, *, extension: str = "") -> str:
        clean = re.sub(r"[^a-f0-9]", "", sha256.lower())
        suffix = f".{extension.strip('.').lower()}" if extension.strip('.') else ""
        return f"sha256/{clean[:2]}/{clean[2:4]}/{clean}{suffix}"

    def put_bytes(
        self,
        data: bytes,
        *,
        media_type: str,
        extension: str = "",
        labels: dict[str, str] | None = None,
    ) -> StoredObject:
        self.root.mkdir(parents=True, exist_ok=True)
        digest = sha256_hex(data)
        relative = self.relative_path(digest, extension=extension)
        target = self.root / relative
        target.parent.mkdir(parents=True, exist_ok=True)
        if not target.exists():
            target.write_bytes(data)
        stored = StoredObject(
            sha256=digest,
            relative_path=relative,
            media_type=media_type,
            size_bytes=len(data),
            created_at=utc_now(),
            labels=labels or {},
        )
        manifest = target.with_suffix(target.suffix + '.json')
        if not manifest.exists():
            manifest.write_text(
                json.dumps(stored.to_json(), indent=2, sort_keys=True) + "\n",
                encoding='utf-8',
            )
        return stored

    def put_text(
        self,
        text: str,
        *,
        media_type: str = 'text/plain',
        extension: str = 'txt',
        labels: dict[str, str] | None = None,
    ) -> StoredObject:
        return self.put_bytes(
            text.encode('utf-8'),
            media_type=media_type,
            extension=extension,
            labels=labels,
        )


@dataclass(frozen=True)
class Episode:
    id: str
    project_id: str
    run_id: str
    request: str
    outcome: str
    summary: str
    failure: str
    lessons: str
    files_changed: tuple[str, ...]
    evidence_hashes: tuple[str, ...]
    completed_items: tuple[str, ...] = ()
    failed_items: tuple[str, ...] = ()
    tags: tuple[str, ...] = ()
    mutations: int = 0
    tool_calls: int = 0
    pinned: bool = False
    completed_at: str = ""


@dataclass(frozen=True)
class MemoryAdmissionDecision:
    status: str
    reason: str
    retrieval_allowed: bool
    candidate_skill: bool
    diagnostic_only: bool


class MemoryAdmissionPolicy:
    def __init__(
        self,
        *,
        require_evidence_hashes: bool = True,
        quarantine_unsuccessful_runs: bool = True,
        promote_successful_mutations: bool = True,
    ) -> None:
        self.require_evidence_hashes = require_evidence_hashes
        self.quarantine_unsuccessful_runs = quarantine_unsuccessful_runs
        self.promote_successful_mutations = promote_successful_mutations

    def evaluate(self, episode: Episode) -> MemoryAdmissionDecision:
        if episode.pinned:
            return MemoryAdmissionDecision('admitted', 'pinned memory remains retrievable', True, False, False)
        if self.require_evidence_hashes and not episode.evidence_hashes:
            return MemoryAdmissionDecision('rejected', 'memory admission requires evidence hashes', False, False, False)
        conversational = self._looks_conversational(episode.request) and not episode.files_changed and not episode.completed_items
        if conversational:
            return MemoryAdmissionDecision('rejected', 'conversational turns do not enter semantic memory', False, False, False)
        if episode.outcome == 'succeeded':
            reusable = bool(episode.files_changed or episode.completed_items) and (episode.mutations > 0 or episode.tool_calls >= 2)
            return MemoryAdmissionDecision(
                'admitted',
                'successful governed run with evidence enters semantic memory',
                True,
                reusable and self.promote_successful_mutations,
                False,
            )
        if self.quarantine_unsuccessful_runs:
            return MemoryAdmissionDecision(
                'quarantined',
                'unsuccessful run is preserved as diagnostic-only memory',
                False,
                False,
                True,
            )
        return MemoryAdmissionDecision('rejected', 'unsuccessful runs are not retained in retrieval memory', False, False, False)

    @staticmethod
    def _looks_conversational(request: str) -> bool:
        lowered = request.lower()
        return any(token in lowered for token in ('what is', 'tell me', 'who is', 'summarize', 'explain'))


@dataclass(frozen=True)
class FreshnessAssessment:
    state: str
    age_days: int
    warning: str


class ResearchFreshnessPolicy:
    def __init__(self, *, fresh_days: int = 30, aging_days: int = 180, citation_required: bool = True) -> None:
        self.fresh_days = fresh_days
        self.aging_days = aging_days
        self.citation_required = citation_required

    def evaluate(self, captured_at: dt.datetime, *, now: dt.datetime | None = None) -> FreshnessAssessment:
        reference = now or dt.datetime.now(dt.timezone.utc)
        age_days = max(0, int((reference - captured_at).total_seconds() // 86400))
        if age_days <= self.fresh_days:
            return FreshnessAssessment('fresh', age_days, 'citation required')
        if age_days <= self.aging_days:
            return FreshnessAssessment('aging', age_days, 'cite capture date when using this source')
        return FreshnessAssessment('stale', age_days, 'stale research requires an explicit warning and citation')


@dataclass(frozen=True)
class SkillCandidate:
    id: str
    project_id: str
    source_episode_id: str
    title: str
    instructions: str
    triggers: tuple[str, ...]
    recommended_tools: tuple[str, ...]
    evidence_hashes: tuple[str, ...]
    candidate_hash: str
    created_at: str

    def to_json(self) -> dict[str, object]:
        return {
            'schemaVersion': '1.0.0',
            'id': self.id,
            'projectId': self.project_id,
            'sourceEpisodeId': self.source_episode_id,
            'title': self.title,
            'instructions': self.instructions,
            'triggers': list(self.triggers),
            'recommendedTools': list(self.recommended_tools),
            'evidenceHashes': list(self.evidence_hashes),
            'candidateHash': self.candidate_hash,
            'createdAt': self.created_at,
        }


@dataclass(frozen=True)
class PublishedSkill:
    id: str
    candidate_id: str
    version: int
    title: str
    instructions: str
    recommended_tools: tuple[str, ...]
    approval_note: str
    manifest_hash: str
    published_at: str

    def to_json(self) -> dict[str, object]:
        return {
            'schemaVersion': '1.0.0',
            'id': self.id,
            'candidateId': self.candidate_id,
            'version': self.version,
            'title': self.title,
            'instructions': self.instructions,
            'recommendedTools': list(self.recommended_tools),
            'approvalNote': self.approval_note,
            'manifestHash': self.manifest_hash,
            'publishedAt': self.published_at,
        }


def extract_skill_candidate(episode: Episode, decision: MemoryAdmissionDecision) -> SkillCandidate | None:
    if not decision.candidate_skill:
        return None
    triggers = tuple(sorted({token for token in _tokenize(episode.request) if token not in {'the', 'and', 'for', 'with', 'from'}}))[:8]
    tools = {'read_file', 'inspect_file', 'verify_project'}
    if episode.mutations > 0:
        tools.add('write_file')
    if episode.tool_calls >= 2:
        tools.add('run_command')
    title = episode.request.strip()[:120] or 'Governed procedure'
    instructions = '\n'.join(
        part for part in (
            f'Objective: {episode.request.strip()}',
            f'Summary: {episode.summary.strip()}' if episode.summary.strip() else '',
            f'Lessons: {episode.lessons.strip()}' if episode.lessons.strip() else '',
            f"Changed files: {', '.join(episode.files_changed)}" if episode.files_changed else '',
        )
        if part
    )
    payload = {
        'projectId': episode.project_id,
        'sourceEpisodeId': episode.id,
        'title': title,
        'instructions': instructions,
        'triggers': list(triggers),
        'recommendedTools': sorted(tools),
        'evidenceHashes': list(episode.evidence_hashes),
    }
    return SkillCandidate(
        id=f'skill_candidate_{sha256_text(episode.id)[:12]}',
        project_id=episode.project_id,
        source_episode_id=episode.id,
        title=title,
        instructions=instructions,
        triggers=triggers,
        recommended_tools=tuple(sorted(tools)),
        evidence_hashes=episode.evidence_hashes,
        candidate_hash=sha256_text(canonical_json(payload)),
        created_at=utc_now(),
    )


def evaluate_skill_candidate(
    candidate: SkillCandidate,
    *,
    replay_passed: bool,
    approval_required: bool = True,
) -> dict[str, object]:
    issues: list[str] = []
    if not candidate.evidence_hashes:
        issues.append('candidate_missing_evidence')
    if any(tool not in ALLOWED_SKILL_TOOLS for tool in candidate.recommended_tools):
        issues.append('candidate_permission_expansion')
    if not replay_passed:
        issues.append('candidate_replay_required')
    if approval_required:
        issues.append('candidate_requires_approval')
    return {
        'candidateId': candidate.id,
        'passed': not issues,
        'issues': issues,
        'evaluationHash': sha256_text(canonical_json({
            'candidateHash': candidate.candidate_hash,
            'replayPassed': replay_passed,
            'approvalRequired': approval_required,
            'issues': issues,
        })),
    }


def publish_skill(
    candidate: SkillCandidate,
    *,
    approval_note: str,
    replay_passed: bool,
    version: int = 1,
) -> PublishedSkill:
    if not replay_passed:
        raise ValueError('replay_passed must be true before publication')
    if not approval_note.strip():
        raise ValueError('approval_note is required')
    if any(tool not in ALLOWED_SKILL_TOOLS for tool in candidate.recommended_tools):
        raise ValueError('candidate expands permissions')
    manifest = {
        'candidateId': candidate.id,
        'title': candidate.title,
        'instructions': candidate.instructions,
        'recommendedTools': list(candidate.recommended_tools),
        'approvalNote': approval_note.strip(),
        'version': version,
    }
    return PublishedSkill(
        id=f'published_skill_{sha256_text(candidate.id + str(version))[:12]}',
        candidate_id=candidate.id,
        version=version,
        title=candidate.title,
        instructions=candidate.instructions,
        recommended_tools=candidate.recommended_tools,
        approval_note=approval_note.strip(),
        manifest_hash=sha256_text(canonical_json(manifest)),
        published_at=utc_now(),
    )

File: lib/product/knowledge_memory_v2.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'crypto_utils.dart';
import 'domain.dart';
import 'repository.dart';

class StoredObjectRecord {
  const StoredObjectRecord({
    required this.sha256,
    required this.relativePath,
    required this.mediaType,
    required this.sizeBytes,
    required this.createdAt,
    this.labels = const <String, String>{},
  });

  final String sha256;
  final String relativePath;
  final String mediaType;
  final int sizeBytes;
  final DateTime createdAt;
  final Map<String, String> labels;

  Map<String, dynamic> toJson() => <String, dynamic>{
    'schemaVersion': '1.0.0',
    'sha256': sha256,
    'relativePath': relativePath,
    'mediaType': mediaType,
    'sizeBytes': sizeBytes,
    'createdAt': createdAt.toUtc().toIso8601String(),
    'labels': labels,
  };

  factory StoredObjectRecord.fromJson(Map<String, dynamic> json) =>
      StoredObjectRecord(
        sha256: json['sha256']?.toString() ?? '',
        relativePath: json['relativePath']?.toString() ?? '',
        mediaType: json['mediaType']?.toString() ?? 'application/octet-stream',
        sizeBytes: int.tryParse(json['sizeBytes']?.toString() ?? '') ?? 0,
        createdAt: parseUtc(json['createdAt'], fallback: DateTime.now()),
        labels: mapValue(
          json['labels'],
        ).map((key, value) => MapEntry(key, value.toString())),
      );
}

class ContentAddressedObjectStore {
  ContentAddressedObjectStore(this.root);

  final Directory root;

  Future<void> initialize() => root.create(recursive: true);

  String relativePathForHash(String sha256, {String extension = ''}) {
    final clean = sha256.replaceAll(RegExp(r'[^a-fA-F0-9]'), '').toLowerCase();
    final safe = clean.length >= 4 ? clean : clean.padRight(4, '0');
    final suffix = extension.trim().isEmpty
        ? ''
        : '.${extension.replaceAll('.', '').toLowerCase()}';
    return 'sha256/${safe.substring(0, 2)}/${safe.substring(2, 4)}/$safe$suffix';
  }

  File fileForRelativePath(String relativePath) {
    final clean = relativePath.replaceAll('\\', '/');
    if (clean.startsWith('/') ||
        clean.split('/').any((segment) => segment.isEmpty || segment == '..')) {
      throw ArgumentError.value(
        relativePath,
        'relativePath',
        'Invalid object-store path.',
      );
    }
    final file = File(
      '${root.path}${Platform.pathSeparator}${clean.replaceAll('/', Platform.pathSeparator)}',
    ).absolute;
    final rootPath = root.absolute.path.replaceAll('\\', '/');
    final candidate = file.path.replaceAll('\\', '/');
    final normalizedRoot = Platform.isWindows
        ? rootPath.toLowerCase()
        : rootPath;
    final normalizedCandidate = Platform.isWindows
        ? candidate.toLowerCase()
        : candidate;
    if (normalizedCandidate != normalizedRoot &&
        !normalizedCandidate.startsWith('$normalizedRoot/')) {
      throw StateError('Object-store path escapes the object root.');
    }
    return file;
  }

  Future<StoredObjectRecord> putBytes(
    List<int> bytes, {
    required String mediaType,
    String extension = '',
    Map<String, String> labels = const <String, String>{},
  }) async {
    await initialize();
    final digest = Sha256.hex(bytes);
    final relative = relativePathForHash(digest, extension: extension);
    final file = fileForRelativePath(relative);
    if (!await file.exists()) {
      await file.parent.create(recursive: true);
      await file.writeAsBytes(bytes, flush: true);
    }
    final manifest = StoredObjectRecord(
      sha256: digest,
      relativePath: relative,
      mediaType: mediaType,
      sizeBytes: bytes.length,
      createdAt: DateTime.now().toUtc(),
      labels: labels,
    );
    final envelope = File('${file.path}.json');
    if (!await envelope.exists()) {
      await envelope.writeAsString(
        '${const JsonEncoder.withIndent('  ').convert(manifest.toJson())}\n',
        flush: true,
      );
    }
    return manifest;
  }

  Future<StoredObjectRecord> putText(
    String text, {
    String mediaType = 'text/plain',
    String extension = 'txt',
    Map<String, String> labels = const <String, String>{},
  }) {
    return putBytes(
      utf8.encode(text),
      mediaType: mediaType,
      extension: extension,
      labels: labels,
    );
  }

  Future<Uint8List> readBytes(String relativePath) async {
    final file = fileForRelativePath(relativePath);
    return Uint8List.fromList(await file.readAsBytes());
  }
}

enum ResearchFreshnessState { fresh, aging, stale, unknown }

class ResearchFreshnessPolicy {
  const ResearchFreshnessPolicy({
    this.freshDays = 30,
    this.agingDays = 180,
    this.citationRequired = true,
    this.staleWarningRequired = true,
  });

  final int freshDays;
  final int agingDays;
  final bool citationRequired;
  final bool staleWarningRequired;

  ResearchFreshnessState evaluate(DateTime capturedAt, {DateTime? now}) {
    final reference = (now ?? DateTime.now().toUtc()).toUtc();
    final ageDays = reference.difference(capturedAt.toUtc()).inDays;
    if (ageDays < 0) {
      return ResearchFreshnessState.unknown;
    }
    if (ageDays <= freshDays) {
      return ResearchFreshnessState.fresh;
    }
    if (ageDays <= agingDays) {
      return ResearchFreshnessState.aging;
    }
    return ResearchFreshnessState.stale;
  }

  String labelFor(DateTime capturedAt, {DateTime? now}) =>
      evaluate(capturedAt, now: now).name;
}

class MemoryAdmissionDecision {
  const MemoryAdmissionDecision({
    required this.status,
    required this.reason,
    required this.retrievalAllowed,
    required this.candidateSkill,
    required this.diagnosticOnly,
  });

  final String status;
  final String reason;
  final bool retrievalAllowed;
  final bool candidateSkill;
  final bool diagnosticOnly;

  Map<String, dynamic> toJson() => <String, dynamic>{
    'status': status,
    'reason': reason,
    'retrievalAllowed': retrievalAllowed,
    'candidateSkill': candidateSkill,
    'diagnosticOnly': diagnosticOnly,
  };
}

class MemoryAdmissionPolicy {
  const MemoryAdmissionPolicy({
    this.requireEvidenceHashes = true,
    this.quarantineUnsuccessfulRuns = true,
    this.promoteSuccessfulMutations = true,
  });

  final bool requireEvidenceHashes;
  final bool quarantineUnsuccessfulRuns;
  final bool promoteSuccessfulMutations;

  MemoryAdmissionDecision evaluateEpisode(MemoryEpisode episode) {
    if (episode.pinned) {
      return const MemoryAdmissionDecision(
        status: 'admitted',
        reason: 'Pinned memory remains available.',
        retrievalAllowed: true,
        candidateSkill: false,
        diagnosticOnly: false,
      );
    }
    if (requireEvidenceHashes && episode.evidenceHashes.isEmpty) {
      return const MemoryAdmissionDecision(
        status: 'rejected',
        reason: 'Memory admission requires evidence hashes.',
        retrievalAllowed: false,
        candidateSkill: false,
        diagnosticOnly: false,
      );
    }
    final conversational =
        _looksConversational(episode.request) &&
        episode.filesChanged.isEmpty &&
        episode.completedItems.isEmpty;
    if (conversational) {
      return const MemoryAdmissionDecision(
        status: 'rejected',
        reason: 'Conversational turns do not enter semantic project memory.',
        retrievalAllowed: false,
        candidateSkill: false,
        diagnosticOnly: false,
      );
    }
    if (episode.outcome == RunState.succeeded) {
      final reusable =
          (episode.filesChanged.isNotEmpty ||
              episode.completedItems.isNotEmpty) &&
          (episode.mutations > 0 || episode.toolCalls >= 2);
      return MemoryAdmissionDecision(
        status: 'admitted',
        reason:
            'Successful governed work with evidence is eligible for retrieval.',
        retrievalAllowed: true,
        candidateSkill: promoteSuccessfulMutations && reusable,
        diagnosticOnly: false,
      );
    }
    if (quarantineUnsuccessfulRuns) {
      return const MemoryAdmissionDecision(
        status: 'quarantined',
        reason: 'Unsuccessful runs are retained only as diagnostic memory.',
        retrievalAllowed: false,
        candidateSkill: false,
        diagnosticOnly: true,
      );
    }
    return const MemoryAdmissionDecision(
      status: 'rejected',
      reason: 'Unsuccessful memory is disabled.',
      retrievalAllowed: false,
      candidateSkill: false,
      diagnosticOnly: false,
    );
  }

  bool _looksConversational(String request) {
    final lower = request.toLowerCase();
    return const <String>[
      'what is',
      'tell me',
      'who is',
      'summarize',
      'explain',
    ].any(lower.contains);
  }
}

class SkillCandidateRecord {
  const SkillCandidateRecord({
    required this.id,
    required this.projectId,
    required this.sourceEpisodeId,
    required this.title,
    required this.instructions,
    required this.triggers,
    required this.recommendedTools,
    required this.evidenceHashes,
    required this.candidateHash,
    required this.createdAt,
  });

  final String id;
  final String projectId;
  final String sourceEpisodeId;
  final String title;
  final String instructions;
  final Set<String> triggers;
  final Set<String> recommendedTools;
  final List<String> evidenceHashes;
  final String candidateHash;
  final DateTime createdAt;

  Map<String, dynamic> toJson() => <String, dynamic>{
    'schemaVersion': '1.0.0',
    'id': id,
    'projectId': projectId,
    'sourceEpisodeId': sourceEpisodeId,
    'title': title,
    'instructions': instructions,
    'triggers': triggers.toList()..sort(),
    'recommendedTools': recommendedTools.toList()..sort(),
    'evidenceHashes': evidenceHashes,
    'candidateHash': candidateHash,
    'createdAt': createdAt.toUtc().toIso8601String(),
  };

  factory SkillCandidateRecord.fromJson(Map<String, dynamic> json) =>
      SkillCandidateRecord(
        id: json['id']?.toString() ?? newId('skill_candidate'),
        projectId: json['projectId']?.toString() ?? '',
        sourceEpisodeId: json['sourceEpisodeId']?.toString() ?? '',
        title: json['title']?.toString() ?? '',
        instructions: json['instructions']?.toString() ?? '',
        triggers: stringList(json['triggers']).toSet(),
        recommendedTools: stringList(json['recommendedTools']).toSet(),
        evidenceHashes: stringList(json['evidenceHashes']),
        candidateHash: json['candidateHash']?.toString() ?? '',
        createdAt: parseUtc(json['createdAt'], fallback: DateTime.now()),
      );
}

class PublishedSkillRecord {
  const PublishedSkillRecord({
    required this.id,
    required this.candidateId,
    required this.version,
    required this.title,
    required this.instructions,
    required this.recommendedTools,
    required this.approvalNote,
    required this.manifestHash,
    required this.publishedAt,
  });

  final String id;
  final String candidateId;
  final int version;
  final String title;
  final String instructions;
  final Set<String> recommendedTools;
  final String approvalNote;
  final String manifestHash;
  final DateTime publishedAt;

  Map<String, dynamic> toJson() => <String, dynamic>{
    'schemaVersion': '1.0.0',
    'id': id,
    'candidateId': candidateId,
    'version': version,
    'title': title,
    'instructions': instructions,
    'recommendedTools': recommendedTools.toList()..sort(),
    'approvalNote': approvalNote,
    'manifestHash': manifestHash,
    'publishedAt': publishedAt.toUtc().toIso8601String(),
  };

  factory PublishedSkillRecord.fromJson(Map<String, dynamic> json) =>
      PublishedSkillRecord(
        id: json['id']?.toString() ?? newId('published_skill'),
        candidateId: json['candidateId']?.toString() ?? '',
        version: int.tryParse(json['version']?.toString() ?? '') ?? 1,
        title: json['title']?.toString() ?? '',
        instructions: json['instructions']?.toString() ?? '',
        recommendedTools: stringList(json['recommendedTools']).toSet(),
        approvalNote: json['approvalNote']?.toString() ?? '',
        manifestHash: json['manifestHash']?.toString() ?? '',
        publishedAt: parseUtc(json['publishedAt'], fallback: DateTime.now()),
      );
}

class SkillPublicationService {
  const SkillPublicationService({
    required this.candidateRepository,
    required this.publishedRepository,
    required this.objectStore,
  });

  final EntityRepository<SkillCandidateRecord> candidateRepository;
  final EntityRepository<PublishedSkillRecord> publishedRepository;
  final ContentAddressedObjectStore objectStore;

  Future<SkillCandidateRecord?> extractFromEpisode(
    MemoryEpisode episode,
  ) async {
    final decision = const MemoryAdmissionPolicy().evaluateEpisode(episode);
    if (!decision.candidateSkill) {
      return null;
    }
    final tokens = _terms(episode.request).take(8).toSet();
    final tools = <String>{'read_file', 'inspect_file', 'verify_project'};
    if (episode.mutations > 0) {
      tools.add('write_file');
    }
    if (episode.toolCalls >= 2) {
      tools.add('run_command');
    }
    final instructions = <String>[
      'Objective: ${episode.request.trim()}',
      if (episode.summary.trim().isNotEmpty)
        'Summary: ${episode.summary.trim()}',
      if (episode.lessons.trim().isNotEmpty)
        'Lessons: ${episode.lessons.trim()}',
      if (episode.filesChanged.isNotEmpty)
        'Changed files: ${episode.filesChanged.join(', ')}',
    ].join('\n');
    final payload = <String, dynamic>{
      'projectId': episode.projectId,
      'sourceEpisodeId': episode.id,
      'title': episode.request.trim(),
      'instructions': instructions,
      'triggers': tokens.toList()..sort(),
      'recommendedTools': tools.toList()..sort(),
      'evidenceHashes': episode.evidenceHashes,
    };
    final candidate = SkillCandidateRecord(
      id: newId('skill_candidate'),
      projectId: episode.projectId,
      sourceEpisodeId: episode.id,
      title: episode.request.trim().isEmpty
          ? 'Governed procedure'
          : episode.request.trim(),
      instructions: instructions,
      triggers: tokens,
      recommendedTools: tools,
      evidenceHashes: episode.evidenceHashes,
      candidateHash: Sha256.text(canonicalJson(payload)),
      createdAt: DateTime.now().toUtc(),
    );
    await candidateRepository.put(candidate);
    await objectStore.putText(
      canonicalJson(candidate.toJson()),
      mediaType: 'application/json',
      extension: 'json',
      labels: const <String, String>{'kind': 'skill-candidate'},
    );
    return candidate;
  }

  Future<List<SkillCandidateRecord>> listCandidates() async {
    final values = await candidateRepository.all();
    values.sort((a, b) => b.createdAt.compareTo(a.createdAt));
    return values;
  }

  Future<List<PublishedSkillRecord>> listPublished() async {
    final values = await publishedRepository.all();
    values.sort((a, b) => b.publishedAt.compareTo(a.publishedAt));
    return values;
  }

  Future<PublishedSkillRecord> publishCandidate(
    SkillCandidateRecord candidate, {
    required String approvalNote,
    required bool replayPassed,
  }) async {
    if (!replayPassed) {
      throw StateError(
        'A candidate skill requires a passing replay evaluation.',
      );
    }
    if (approvalNote.trim().isEmpty) {
      throw ArgumentError.value(
        approvalNote,
        'approvalNote',
        'Publishing a skill requires an explicit approval note.',
      );
    }
    final existing = (await publishedRepository.all())
        .where((item) => item.candidateId == candidate.id)
        .toList();
    final version = existing.length + 1;
    final manifest = <String, dynamic>{
      'candidateId': candidate.id,
      'title': candidate.title,
      'instructions': candidate.instructions,
      'recommendedTools': candidate.recommendedTools.toList()..sort(),
      'approvalNote': approvalNote.trim(),
      'version': version,
    };
    final published = PublishedSkillRecord(
      id: newId('published_skill'),
      candidateId: candidate.id,
      version: version,
      title: candidate.title,
      instructions: candidate.instructions,
      recommendedTools: candidate.recommendedTools,
      approvalNote: approvalNote.trim(),
      manifestHash: Sha256.text(canonicalJson(manifest)),
      publishedAt: DateTime.now().toUtc(),
    );
    await publishedRepository.put(published);
    await objectStore.putText(
      canonicalJson(published.toJson()),
      mediaType: 'application/json',
      extension: 'json',
      labels: const <String, String>{'kind': 'published-skill'},
    );
    return published;
  }

  Iterable<String> _terms(String value) sync* {
    final matches = RegExp(r'[a-z0-9_./-]+').allMatches(value.toLowerCase());
    for (final match in matches) {
      final token = match.group(0);
      if (token != null && token.isNotEmpty) {
        yield token;
      }
    }
  }
}
